### PR TITLE
fix the logic of isAuthorized and Authorized

### PR DESCRIPTION
**Pull request by amine-benselim**
this PR was meant to address a bug we found on Customer dataFences but eventually turned out to be an issue we have with dataFences and permissions in general. 

previously, if in the code we demand dataFences AND general permissions, we checked first for the dataFences, and if the user turned out not to have those permissions, we return early with `isAuthorized = false` WITHOUT checking the general permissions (and action rights).

now, we first check dataFences and THEN we also check for the general permissions if they are met. 

I also fixed the tests that used deprecated params (mainly to silence Typescript 🤷‍♂)

### CODE REVIEW

**Comment by amine-benselim** [comment]
`yarn typecheck` does not fail for me locally so I don't know what those TS errors are 🤔

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 1]
```diff
@@ -76,39 +76,32 @@ const defaultProps: Pick<Props, 'shouldMatchSomePermissions'> = {
...
+  if (props.demandedDataFences && props.demandedDataFences.length > 0) {
     if (!props.selectDataFenceData) {
       reportErrorToSentry(
         new Error(
           `@commercetools-frontend/permissions/Authorized: Missing data fences selector "selectDataFenceData".`
         )
       );
-      return <React.Fragment>{props.render(false)}</React.Fragment>;
     }
-    return (
```
Comment (amine-benselim) [inline packages/permissions/src/components/authorized/authorized.tsx]: it didnt make sense to return here because the user might meet the general permissions without meeting the dataFence permissions

**Comment by emmenko** [comment]
> `yarn typecheck` does not fail for me locally so I don't know what those TS errors are

Hmm you might need to remove your `dist` folders

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 2]
```diff
@@ -122,18 +115,21 @@ const Authorized = (props: Props) => {
...
+  hasDemandedActionRights = hasEveryActionRight(
     props.demandedActionRights || [],
     props.actualActionRights
   );
 
   return (
     <React.Fragment>
-      {props.render(hasDemandedPermissions && hasDemandedActionRights)}
+      {props.render(
+        hasDemandedDataFences ||
```
Comment (amine-benselim) [inline packages/permissions/src/components/authorized/authorized.tsx]: this is where we fallback to general permissions if dataFence permissions are not met.

File: packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx [hunk 3]
```diff
@@ -55,15 +55,13 @@ const renderWithPermissions = (demandedPermissions: string[]) => {
         suspension: {
           isActive: true,
         },
-        allAppliedPermissions: [{ name: 'canViewProducts', value: true }],
-        allAppliedActionRights: [
-          {
-            group: 'products',
-            name: 'canEditPrices',
-            value: false,
+        permissions: { canViewProducts: true },
```
Comment (emmenko) [inline packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]: Why did you change those? 🤔

Comment (emmenko) [reply packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]: They need to stay as they were

File: packages/permissions/src/utils/has-permissions.ts [hunk 4]
```diff
@@ -238,18 +244,15 @@ const getDataFenceByPermissionGroup = (
 
 export const hasAppliedDataFence = (options: TOptionsForAppliedDataFence) => {
   // If user has `*Manage` as their `appliedPermission`, we bypass `dataFence`.
-  const hasAppliedManagePermission = options.demandedPermissions.some(
-    demandedPermission => {
-      if (options.actualPermissions) {
```
Comment (amine-benselim) [inline packages/permissions/src/utils/has-permissions.ts]: I removed the check from here and put it inside `hasExactPermission`

File: packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx (hunk 3 above)
Comment (amine-benselim) [reply packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]: I thought those were deprecated. I changed the tests to reflect the new params (permissions, dataFences, actionRights). also, typescript was complaining about them inside the test...

Comment (emmenko) [reply packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]: I'm not sure what you mean with "deprecated" but this is what we get from the graphql query

**Comment by amine-benselim** [comment]
> > `yarn typecheck` does not fail for me locally so I don't know what those TS errors are
> 
> Hmm you might need to remove your `dist` folders

deleting the `dist` folders actually revealed the errors. why is that so? 🤔

File: packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx (hunk 3 above)
Comment (amine-benselim) [reply packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]:
```
// For backwards compatibility we need to denormalize the given `permissions` option
// (which is now deprecated) to `allAppliedPermissions`, in order to pass the value
// to the `project` prop in the application context provider.
```

I got this backwards 😓

Comment (amine-benselim) [reply packages/permissions/src/components/branch-on-permissions/branch-on-permissions.spec.tsx]: reverted 🙏

**Comment by emmenko** [comment]
> deleting the `dist` folders actually revealed the errors. why is that so?

I suppose because of the type declarations in the packages that get resolved from the `dist` entry point

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 5]
```diff
@@ -122,18 +115,21 @@ const Authorized = (props: Props) => {
       { extra: namesOfNonConfiguredPermissions }
     );
 
-  const hasDemandedPermissions = props.shouldMatchSomePermissions
```
Comment (emmenko) [inline packages/permissions/src/components/authorized/authorized.tsx]: This can stay as `const`, right?

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 6]
```diff
@@ -122,18 +115,21 @@ const Authorized = (props: Props) => {
...
+  hasDemandedPermissions = props.shouldMatchSomePermissions
     ? hasSomePermissions(props.demandedPermissions, props.actualPermissions)
     : hasEveryPermissions(props.demandedPermissions, props.actualPermissions);
 
-  const hasDemandedActionRights = hasEveryActionRight(
```
Comment (emmenko) [inline packages/permissions/src/components/authorized/authorized.tsx]: Same here

File: packages/permissions/src/components/authorized/authorized.spec.tsx [hunk 7]
```diff
@@ -283,8 +283,8 @@ describe('rendering', () => {
             });
             shallow(<Authorized {...props} />);
           });
-          it('should indicate as not authorized', () => {
-            expect(props.render).toHaveBeenCalledWith(false);
+          it('should indicate as authorized', () => {
+            expect(props.render).toHaveBeenCalledWith(true);
```
Comment (emmenko) [inline packages/permissions/src/components/authorized/authorized.spec.tsx]: Just double checking: is this the expected outcome now?

Comment (amine-benselim) [inline packages/permissions/src/components/authorized/authorized.spec.tsx]: this test case is exactly what was wrong about our logic, 
if you look at the actual and demanded permissions: 
```
demandedPermissions: ['ViewOrders'],
actualPermissions: {	        
  canViewOrders: true,	           
  canManageOrders: false,	               
},
```
the user IS authorized even if he did not have a manage orders DataFence, because in the code we also demanded the general permission `ViewOrders` and the user meets this permission.

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 8]
```diff
@@ -76,64 +75,47 @@ const defaultProps: Pick<Props, 'shouldMatchSomePermissions'> = {
...
+      actualPermissions: props.actualPermissions,
+      demandedPermissions: props.demandedPermissions,
+    });
   }
 
-  // If no data fences have been provided, fall back to normal permissions + action rights.
-  if (!props.actualPermissions)
-    return <React.Fragment>{props.render(false)}</React.Fragment>;
-
-  const namesOfNonConfiguredPermissions = getInvalidPermissions(
```
Comment (emmenko) [inline packages/permissions/src/components/authorized/authorized.tsx]: Can you explain again why we don't need this anymore?

File: packages/permissions/src/utils/has-permissions.spec.tsx [hunk 9]
```diff
@@ -185,23 +184,3 @@ describe('hasEveryActionRight', () => {
     ).toBe(false);
   });
 });
-
-describe('getInvalidPermissions', () => {
```
Comment (tdeekens) [inline packages/permissions/src/utils/has-permissions.spec.tsx]: This feels like an unrelated change.

File: packages/permissions/src/utils/has-permissions.ts [hunk 10]
```diff
@@ -60,10 +59,10 @@ type TActualDataFence = {
 
 type TOptionsForAppliedDataFence = {
   demandedDataFences: TDemandedDataFence[];
-  actualDataFences: TDataFences;
-  selectDataFenceData: TSelectDataFenceData;
+  actualDataFences: TDataFences | null;
+  selectDataFenceData?: TSelectDataFenceData;
   actualPermissions: TPermissions | null;
-  demandedPermissions: TPermissionName[];
+  demandedPermissions: TPermissionName[] | null;
```
Comment (tdeekens) [inline packages/permissions/src/utils/has-permissions.ts]: Why did the type signaure have to change?

File: packages/permissions/src/utils/has-permissions.ts [hunk 11]
```diff
@@ -175,27 +175,13 @@ export const hasSomePermissions = (
...
-  );
-};
-
 const hasDemandedDataFence = (options: {
   actualDataFence: TActualDataFence;
   demandedDataFence: TDemandedDataFence;
-  selectDataFenceData: TSelectDataFenceData;
+  selectDataFenceData?: TSelectDataFenceData;
 }): boolean => {
+  if (!options.selectDataFenceData) return false;
```
Comment (tdeekens) [inline packages/permissions/src/utils/has-permissions.ts]: Would you log in this case too?

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 12]
```diff
@@ -76,64 +75,47 @@ const defaultProps: Pick<Props, 'shouldMatchSomePermissions'> = {
...
-  if (namesOfNonConfiguredPermissions.length > 0)
-    reportErrorToSentry(
-      new Error(
-        `@commercetools-frontend/permissions/Authorized: Invalid prop "demandedPermissions" supplied. The permission(s) ${namesOfNonConfiguredPermissions.toString()} is/are not configured through "actualPermissions".`
-      ),
-      { extra: namesOfNonConfiguredPermissions }
-    );
-
-  const hasDemandedPermissions = props.shouldMatchSomePermissions
+  hasDemandedPermissions = props.shouldMatchSomePermissions
```
Comment (tdeekens) [inline packages/permissions/src/components/authorized/authorized.tsx]: This feels wrong. Why initialise it to a boolean argument?

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 13]
```diff
@@ -76,64 +75,47 @@ const defaultProps: Pick<Props, 'shouldMatchSomePermissions'> = {
...
 
 const Authorized = (props: Props) => {
-  // Check first for data fences
-  if (
-    props.actualDataFences &&
-    props.demandedDataFences &&
-    props.demandedDataFences.length > 0
-  ) {
+  // if the user has no permissions and no dataFences assigned to them, they are not authorized
+  if (!props.actualDataFences && !props.actualPermissions) {
```
Comment (adnasa) [inline packages/permissions/src/components/authorized/authorized.tsx]: this case won't happen since `actualDataFences` and `actualPermissions` won't be `undefined`, right?

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.ts [hunk 14]
```diff
@@ -109,7 +110,9 @@ const useIsAuthorized = ({
     actualActionRights
   );
 
-  return hasDemandedPermissions && hasDemandedActionRights;
+  return (
+    hasDemandeDataFences || (hasDemandedPermissions && hasDemandedActionRights)
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.ts]: okay, I see that we lift up the fallback

File: packages/permissions/src/components/authorized/authorized.spec.tsx (hunk 7 above)
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.spec.tsx]: yes

**Comment by amine-benselim** [comment]
I admit I have went too far with the changes, I will revert the change not related to the bug we're facing and then I will open another PR to explore simplifying the code.

File: packages/permissions/src/components/authorized/authorized.tsx (hunk 13 above)
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: actualDataFences can be, and if we look at the old code we where doing this: 
```
if (!props.actualPermissions)	
    return <React.Fragment>{props.render(false)}</React.Fragment>;
```

so also actual permissions can be `null`

File: packages/permissions/src/components/authorized/authorized.tsx (hunk 8 above)
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: I can explain this, but I will do so in another PR 👍

File: packages/permissions/src/components/authorized/authorized.tsx (hunk 12 above)
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: I don't understand, what do you mean exactly? 🤔

File: packages/permissions/src/utils/has-permissions.ts (hunk 11 above)
Comment (amine-benselim) [reply packages/permissions/src/utils/has-permissions.ts]: we log already in the Component level (Authorized and useIsAuthorized).

File: packages/permissions/src/components/restricted-by-permissions/restricted-by-permissions.tsx [hunk 15]
```diff
@@ -55,7 +55,7 @@ const RestrictedByPermissions = (props: Props) => {
       render={applicationContext => (
         <Authorized
           shouldMatchSomePermissions={props.shouldMatchSomePermissions}
-          actualPermissions={applicationContext.permissions}
+          actualPermissions={applicationContext.permissions || {}}
```
Comment (tdeekens) [inline packages/permissions/src/components/restricted-by-permissions/restricted-by-permissions.tsx]: Can you explain why this needs to change?

Comment (amine-benselim) [reply packages/permissions/src/components/restricted-by-permissions/restricted-by-permissions.tsx]: I reverted this change 👍

File: packages/permissions/src/components/authorized/authorized.tsx (hunk 13 above)
Comment (adnasa) [reply packages/permissions/src/components/authorized/authorized.tsx]: 👍 thanks

**Comment by tdeekens** [comment]
Sorry, I fail to follow at all. What is "state of inputs" or "on state" mean here?

**Comment by adnasa** [comment]
state of input: whether a field should be disabled or not.

**Comment by tdeekens** [comment]
@adnasa thanks for clarifying. I think however we need to separate two classes of problems:

1. We architecture doesn't allow us to achieve what we have to (component, hook, render props)
2. This PR doesn't fullfil all cases of UATs

What you're saying sounds like 1. but I think it's more like 2. So we just need to iterate on this and cover all cases and have a test each.

**Comment by amine-benselim** [comment]
I have allowed myself to refactor the permission package to have `Authorized` use `useIsAuthorized` under the hook as the only source of logic. this makes debugging and trying to fix issues simpler. 

I have also moved the tests from `Authorized` to `useIsAuthorized` and they all passed 🎉 

regarding the logic itself, I think the solution we have so far is sound and sufficient. I will discuss this with Adnan in person and will add more test cases for the scenarios we want to cover.

**Comment by amine-benselim** [comment]
@emmenko if you take a look at the Typescript error, you see that it's complaining about `{ __typename: 'StoreDataFence' }` 

the reason for that is because when taking a look at `mc.ts` you notice that this typename is the only one that is not optional (without `?`). I know that the types are generated (althought I don't know where from exactly), so I have no idea why this typename is an exception

File: packages/permissions/src/utils/has-permissions.ts [hunk 16]
```diff
@@ -237,19 +240,6 @@ const getDataFenceByPermissionGroup = (
 };
 
 export const hasAppliedDataFence = (options: TOptionsForAppliedDataFence) => {
-  // If user has `*Manage` as their `appliedPermission`, we bypass `dataFence`.
-  const hasAppliedManagePermission = options.demandedPermissions.some(
```
Comment (amine-benselim) [inline packages/permissions/src/utils/has-permissions.ts]: because we fall back now to general permissions if dataFence permissions are not met, we do not need to check if the use has `manage` general permission first INSIDE `hasAppliedDataFence`, this simplifies the logic further and makes the flow for permission check straight forward: 

1- check if the user meets the dataFence permissions
2- if not, check if he meets the general permissions + action rights

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 17]
```diff
@@ -18,7 +18,6 @@ type TAllAppliedMenuVisibility = {
   value: boolean;
 };
 type TAllAppliedDataFence = {
-  __typename: 'StoreDataFence';
```
Comment (emmenko) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: Can you explain why this is removed? 🤔

Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: because that 'fixes' the Typescript error I told you about in the previous comment, but does not address the original issue which is that only `__typename: 'StoreDataFence'` is not optional in `mc.ts`

Comment (emmenko) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: But what was the original error?

Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: there was no original error. when I was adding more test cases I had to add `__typename: 'StoreDataFence'` to all of my applied dataFences which I think is not right.

Comment (emmenko) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: You mean in the tests? I would define it where is needed to be honest, if that's how the types are. The types are generated from the graphql queries -> schema.

Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: done that 👍

Comment (emmenko) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: Thanks. Btw, if you want to be less repetitive in tests, you can simply have a `createTestStore` function or something which has some defaults values 😉

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 18]
```diff
@@ -161,24 +73,16 @@ const injectAuthorized = <
   Component: React.ComponentType<OwnProps>
 ): React.ComponentType<OwnProps & InjectedProps> => {
   const WrappedComponent = (props: OwnProps) => (
-    <ApplicationContext
```
Comment (adnasa) [inline packages/permissions/src/components/authorized/authorized.tsx]: nice

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 19]
```diff
@@ -187,4 +208,271 @@ describe('when only one permission matches', () => {
...
+            demandedActionRights: [
+              { group: 'products', name: 'PublishProducts' },
+            ],
+            shouldMatchSomePermissions: true,
+          });
+
+          expect(queryByText('Is authorized: No')).toBeInTheDocument();
+        });
+      });
+      describe('if can edit prices on products group', () => {
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: This is a good test.
This proves that it is on a consumer side to be careful when demanding permissions.

because one can expect `isAuthorized=true` if `actualPermissions.ViewProducts=true` and `actualActionRights.canEditPrices=true`, then you get `isAuthorized=true` if you demand this.

However if this happens, we may have a flaw in how we save action rights.

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 20]
```diff
@@ -187,4 +208,271 @@ describe('when only one permission matches', () => {
...
+          expect(queryByText('Is authorized: No')).toBeInTheDocument();
+        });
+      });
+    });
+  });
+
+  describe('dataFences', () => {
+    describe('with general permission', () => {
+      describe('when general permissions is manage', () => {
+        it('should indicate as authorized', () => {
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: should add another describe `when demanded general permissions is...`

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 21]
```diff
@@ -187,4 +208,271 @@ describe('when only one permission matches', () => {
...
+
+          expect(queryByText('Is authorized: Yes')).toBeInTheDocument();
+        });
+      });
+      describe('when general permissions is view', () => {
+        describe('when data fence permission is manage', () => {
+          it('should indicate as authorized', () => {
+            const { queryByText } = testRender({
+              shouldMatchSomePermissions: true,
+              allAppliedPermissions: [{ name: 'canManageOrders', value: true }],
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: but the describe says `when general permissions is view`

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx (hunk 17 above)
Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: I already had that 👍 I could add more abstraction but the test cases require me to change the demanded and actual dataFences and permissions.

**Comment by amine-benselim** [comment]
> it is clearer now with the test.
> We could go write it to be a bit more verbose and make the language clearer
> 
> example
> 
> ```js
> describe('with general permissions view', () => {
>   describe('with data fence permissions view', () => {
>     describe('when demanding data fence manage', () => {
>       it('should...', () => {
>         
>       })
>     })
>   })
>   describe('with data fence permissions manage', () => {
>     describe('when demanding data fence manage', () => {
>       it('should...', () => {
> 
>       })
>     })
>   })
> })
> ```

thanks @adnasa for the suggestion, I already rewrote the tests before reading your comments, can you check them now if they make sense or should I change them even further 🤔

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx (hunk 21 above)
Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: this is refactored now.

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 22]
```diff
@@ -130,86 +130,90 @@ const testRender = ({
...
-            group: 'products',
-            name: 'canEditPrices',
-            value: true,
-          },
-        ],
-        demandedPermissions: ['ManageCustomers'],
-        demandedActionRights: [{ group: 'products', name: 'PublishProducts' }],
-        shouldMatchSomePermissions: false,
-      });
+    describe('when it should not match some permission', () => {
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: I don't get this part.. does this refer to the outcome or criteria?
if it is the criteria, then `when shouldMatchSomePermissions=false`

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 23]
```diff
@@ -436,11 +469,10 @@ describe('when only one permission matches', () => {
           expect(queryByText('Is authorized: Yes')).toBeInTheDocument();
         });
       });
-
-      describe('has demanded dataFence to VIEW or MANAGE orders on the specific store', () => {
-        it('should indicate as authorized', () => {
+      describe('when dataFence permission is not met', () => {
```
Comment (adnasa) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: I don't know. I prefer the language of `when general permissions...`, `when datafence permissions` vs `when demanded datafence permissions ...`

Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: I mean there is a difference between the demanded and the actual and we need to make that difference clear in our test cases.

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx [hunk 24]
```diff
@@ -103,88 +122,431 @@ const testRender = ({
...
-        demandedPermissions: ['ManageCustomers', 'ManageOrders'],
-        allAppliedPermissions: [
-          {
-            name: 'canManageCustomers',
-            value: true,
-          },
-        ],
-        shouldMatchSomePermissions: true,
+  describe('General Permissions', () => {
+    describe('when "shouldMatchSomePermissions=true"', () => {
```
Comment (tdeekens) [inline packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: Nit: I wouldn't even bind the name of the spec the prop name. It's just "describe(`when all permissions should be matched`"

File: packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx (hunk 23 above)
Comment (amine-benselim) [reply packages/permissions/src/hooks/use-is-authorized/use-is-authorized.spec.tsx]: the problem is that it does not make sense to talk about general permissions being view or manage anymore, the dataFence permissions are agnostic to what the general permission are, so in the context of data fences we talk only about wether the user meets the general permissions or not. 

overall, I think the test cases and their structure is good enough for now and would like to get this PR through.

File: packages/permissions/src/components/authorized/authorized.tsx [hunk 25]
```diff
@@ -76,66 +39,15 @@ const defaultProps: Pick<Props, 'shouldMatchSomePermissions'> = {
...
-    props.demandedActionRights || [],
-    props.actualActionRights
-  );
-
-  return (
-    <React.Fragment>
-      {props.render(hasDemandedPermissions && hasDemandedActionRights)}
-    </React.Fragment>
-  );
+  return <React.Fragment>{props.render(isAuthorized)}</React.Fragment>;
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/permissions/src/components/authorized/authorized.tsx]: Yes otherwise TS cannot recognize it as a react component

**Comment by adnasa** [comment]
thanks for this, @amine-benselim 😉

File: packages/permissions/src/components/authorized/authorized.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: I think we don't, I don't know why it was the case before and I just did not change it 
I will remove it now.

[the remark this answers is not in the document]
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: maybe to allow the render prop to return more than one ReactNode? 🤔

[the remark this answers is not in the document]
Comment (amine-benselim) [reply packages/permissions/src/components/authorized/authorized.tsx]: Typescript complained a lot when I deleted it, so I guess I have to keep it 😅
